design/baremetal-operator/raid-api.md

```diff
@@ -0,0 +1,198 @@
+metal3 could do that work for them.
+
+### Goals
+
+1. Provide an API to specify the RAID storage configuration to use
+   when provisioning a host.
+2. Outline the controller/provisioner logic changes needed to pass the
+   instructions to Ironic to have the storage configured.
+3. Support basic RAID levels `0`, `1`, and `1+0`.
+
+### Non-Goals
+
+1. Provide default RAID configuration.
+2. Retain existing RAID configuration on a host if it matches the
+   desired configuration at the time of provisioning (we make no
+   assumptions about preserving the state of the storage system or its
+   contents).
+3. Re-provision a host if the RAID settings change. (Users are
+   expected to explicitly deprovision the host and then reprovision
+   it.)
+4. Support every possible RAID configuration.
+
+## Proposal
+
+### User Stories
+
+1. As a user, I can specify the RAID configuration for a host before
+   provisioning it.
+2. As a user, I can choose between hardware and software RAID for each
+   host.
+
+## Design Details
+
+### API Changes
+
+The `BareMetalHost` API will change to accept RAID configuration
+inputs and to reflect the settings that were used when a host is
+provisioned (in case they are changed after provisioning).
+
+New `spec` fields will be added in a section called `raid`, with
+separate fields describing the hardware and software RAID settings. If
+the hardware settings are provided, the software settings are ignored.
+
+For hardware RAID, the inputs describe how to combine physical disks
+into a RAID volume via the RAID controller hardware in the host.
+
+```yaml
+spec:
+  raid:
+    hardwareVolumes:
+      - sizeGibibytes: 1000
+        level: 1
+        name: volume-name
+        rotational: true
+        numberOfPhysicalDisks: 2
+```
+
+- The RAID level for a volume is required and should be expressed as
+  one of `"1"`, `"0"`, or `"1+0"`.
+- The size for a volume is optional and should be expressed in
+  gibibytes.
+- The number of disks to use for the volume defaults to a minimum
+  value based on the RAID level.
+
+For software RAID, the inputs describe how to configure the operating
+system to see physical disks in a RAID configuration through software
+drivers, without using any special hardware.
+
+```yaml
+spec:
+  raid:
+    softwareVolumes:
+      - sizeGibibytes: 1000
+        level: 1
+        physicalDisks:
+          - deviceName: /dev/sda
+          - deviceName: /dev/sdb
+```
+
+- The RAID level for a volume is required and should be expressed as
+  one of `"1"`, `"0"`, or `"1+0"`.
+- The size for a volume is optional and should be expressed in
+  gibibytes.
+- The disks to use are specified using the same syntax as the
+  `rootDeviceHints`.
+
+Software RAID is constrained in other ways:
+
+- There must be 0, 1, or 2 software RAID devices.
+- The first software RAID volume must use RAID level 1 to ensure the
+  host boots in the event of a disk failure.
```

Comment:
Specifying the RAID level for the root volume seems like a recommendation we could make (changing "must" to "should") but not something we should really be requiring. Does something actually enforce that rule? I don't see it in the current PR.

Reply:
Sorry, i forgot to check the root volume. Now, I added it.

Reply:
I copied this from the comments in the implementation pull request, but I don't think metal3 should be enforcing that rule at all. Is there is a technical reason for it? Would deployment fail if we did not enforce this rule?

Reply:
I'm not sure if this rule exists in Ironic as well, from the docs, it "has to be" RAID-1 in interest of not having a failed boot. and I agree, this could be documented as a "recommendation" rather than a hardcore requirement.

Reply:
I did not find any enforcement measures in the source code of ironic:

[raid.py](https://github.com/openstack/ironic/blob/master/ironic/common/raid.py)
```
def _check_and_return_root_volumes(raid_config):
    """Returns root logical disks after validating RAID config.
    This method checks if multiple logical disks had 'is_root_volume'
    set to True and raises an exception if it is True. Otherwise,
    returns the root logical disk mentioned in the RAID config.
    :param raid_config: target RAID configuration or current RAID
        configuration.
    :returns: the dictionary for the root logical disk if it is
        present, otherwise None.
    :raises: InvalidParameterValue, if there were more than one
        root volume specified in the RAID configuration.
    """
    logical_disks = raid_config['logical_disks']
    root_logical_disks = [x for x in logical_disks if x.get('is_root_volume')]
    if len(root_logical_disks) > 1:
        msg = _("Raid config cannot have more than one root volume. "
                "%d root volumes were specified") % len(root_logical_disks)
        raise exception.InvalidParameterValue(msg)

    if root_logical_disks:
        return root_logical_disks[0]

def validate_configuration(raid_config, raid_config_schema):
    """Validates the RAID configuration passed using JSON schema.
    This method validates a RAID configuration against a RAID configuration
    schema.
    :param raid_config: A dictionary containing RAID configuration information
    :param raid_config_schema: A dictionary which is the schema to be used for
        validation.
    :raises: InvalidParameterValue, if validation of the RAID configuration
        fails.
    """
    try:
        jsonschema.validate(raid_config, raid_config_schema)
    except json_schema_exc.ValidationError as e:
        # NOTE: Even though e.message is deprecated in general, it is said
        # in jsonschema documentation to use this still.
        msg = _("RAID config validation error: %s") % e.message
        raise exception.InvalidParameterValue(msg)

    # Check if there are multiple root volumes specified.
    _check_and_return_root_volumes(raid_config)
```
[raid_config_schema.json](https://github.com/openstack/ironic/blob/master/ironic/drivers/raid_config_schema.json)
```

{
    "$schema": "http://json-schema.org/draft-04/schema#",
    "title": "raid configuration json schema",
    "type": "object",
    "properties": {
        "logical_disks": {
            "type": "array",
            "items": {
                "type": "object",
                "properties": {
                    "raid_level": {
                        "type": "string",
                        "enum": [ "JBOD", "0", "1", "2", "5", "6", "1+0", "5+0", "6+0" ],
                        "description": "RAID level for the logical disk. Valid values are 'JBOD', '0', '1', '2', '5', '6', '1+0', '5+0' and '6+0'. Required."
                    },
                    ......
                    ......
                    ......
                },
                "required": ["raid_level", "size_gb"],
                "additionalProperties": false,
                "dependencies": {
                    "physical_disks": ["controller"]
                }
            },
            "minItems": 1
        }
    },
    "required": ["logical_disks"],
    "additionalProperties": false
}
```

But the description from [docs.openstack.org](https://docs.openstack.org/ironic/latest/admin/raid.html#software-raid) clearly states this:
>Building Linux software RAID in-band (via the Ironic Python Agent ramdisk) is supported starting with the Train release. It is requested by using the agent RAID interface and RAID configuration with all controllers set to software. You can find a software RAID configuration example in Examples for target_raid_config.
>
>There are certain limitations to be aware of:
>- Only the mandatory properties (plus the required controller property) from Target RAID configuration are currently supported.
>- The number of created Software RAID devices must be 1 or 2. If there is only one Software RAID device, it has to be a RAID-1. If there are two, the first one has to be a RAID-1, while the RAID level for the second one can 0, 1, or 1+0. As the first RAID device will be the deployment device, enforcing a RAID-1 reduces the risk of ending up with a non-booting node in case of a disk failure.
>- Building RAID will fail if the target disks are already partitioned. Wipe the disks using e.g. the erase_devices_metadata clean step before building RAID:

Reply:
There are use cases where it wouldn't be suggestive to enforce RAID-1, for instance if the node only has 1 disk on it. Additionally, root disks, in most environments these days only carry the operating system, and no persistent data, in the effort to ensure overall system stability in case there is node failure. That will be another reason to not require a RAID-1 enforcement, maybe Ironic does that for a reason.

Reply:
@dtantsur do you have any insight into this?

Reply:
The ironic `idrac` hardware type/driver's **hardware** RAID does not enforce `RAID-1` for `logical_disks` which are root volumes.

Reply:
That sounds good!

Reply:
Is it safe to remove this requirement, then?

Reply:
I believe so, since it isn't a hard requirement for every driver. It could either change into a ``note`` perhaps? Saying something like "Check if your driver enforces RAID-1 for the first volume" or something akin to it.

Reply:
One of the goals of metal3 is that users shouldn't need to go read the Ironic documentation in order to understand how to use it. So, if there is some driver that does require RAID-1 for root volumes we should document that in the metal3 docs as well. Do we think there is a driver with that requirement? Or are we just not sure that there isn't a driver like that?

Does the difference in driver behavior within Ironic mean that we need some driver-specific validation in metal3?

Reply:
Well, the documentation in Ironic says that it _should_ be RAID-1 (which looks like a recommendation to me). And @rpioso just pointed out that it is not _necessary_ in case of ``idrac``. This appears to me more like a recommendation/usage intricacy than a hardline that needs to be validated in code. 

I agree with the part about documenting this specifically in metal3, one way or the other.

Reply:
I think `software raid` is supported by `ironic_python_agent`, and found the requirements for `software raid`.

[hardware.py](https://github.com/openstack/ironic-python-agent/blob/ddbba070216233e695f90095334ab7ddaef563f2/ironic_python_agent/hardware.py#L1815-L1901)

```python
    def validate_configuration(self, raid_config, node):
        """Validate a (software) RAID configuration
        Validate a given raid_config, in particular with respect to
        the limitations of the current implementation of software
        RAID support.
        :param raid_config: The current RAID configuration in the usual format.
        """
        LOG.debug("Validating Software RAID config: {}".format(raid_config))

        if not raid_config:
            LOG.error("No RAID config passed")
            return False

        logical_disks = raid_config.get('logical_disks')
        if not logical_disks:
            msg = "RAID config contains no logical disks"
            raise errors.SoftwareRAIDError(msg)

        raid_errors = []

        # Only one or two RAID devices are supported for now.
        if len(logical_disks) not in [1, 2]:
            msg = ("Software RAID configuration requires one or "
                   "two logical disks")
            raid_errors.append(msg)

        # All disks need to be flagged for Software RAID
        for logical_disk in logical_disks:
            if logical_disk.get('controller') != 'software':
                msg = ("Software RAID configuration requires all logical "
                       "disks to have 'controller'='software'")
                raid_errors.append(msg)

            physical_disks = logical_disk.get('physical_disks')
            if physical_disks is not None:
                if (not isinstance(physical_disks, list)
                        or len(physical_disks) < 2):
                    msg = ("The physical_disks parameter for software RAID "
                           "must be a list with at least 2 items, each "
                           "specifying a disk in the device hints format")
                    raid_errors.append(msg)
                if any(not isinstance(item, dict) for item in physical_disks):
                    msg = ("The physical_disks parameter for software RAID "
                           "must be a list of device hints (dictionaries)")
                    raid_errors.append(msg)

        # The first RAID device needs to be RAID-1.
        if logical_disks[0]['raid_level'] != '1':
            msg = ("Software RAID Configuration requires RAID-1 for the "
                   "first logical disk")
            raid_errors.append(msg)

        # Additional checks when we have two RAID devices.
        if len(logical_disks) == 2:
            size1 = logical_disks[0]['size_gb']
            size2 = logical_disks[1]['size_gb']

            # Only one logical disk is allowed to span the whole device.
            if size1 == 'MAX' and size2 == 'MAX':
                msg = ("Software RAID can have only one RAID device with "
                       "size 'MAX'")
                raid_errors.append(msg)

            # Check the accepted RAID levels.
            current_level = logical_disks[1]['raid_level']
            if current_level not in SUPPORTED_SOFTWARE_RAID_LEVELS:
                msg = ("Software RAID configuration does not support "
                       "RAID level %s" % current_level)
                raid_errors.append(msg)
            physical_device_count = len(self.list_block_devices())
            if current_level == '5' and physical_device_count < 3:
                msg = ("Software RAID configuration is not possible for "
                       "RAID level 5 with only %s block devices found."
                       % physical_device_count)
                raid_errors.append(msg)
            if current_level == '6' and physical_device_count < 4:
                msg = ("Software RAID configuration is not possible for "
                       "RAID level 6 with only %s block devices found."
                       % physical_device_count)
                raid_errors.append(msg)
        if raid_errors:
            error = ('Could not validate Software RAID config for %(node)s: '
                     '%(errors)s') % {'node': node['uuid'],
                                      'errors': '; '.join(raid_errors)}
            raise errors.SoftwareRAIDError(error)
```

Reply:
It does say that "every logical disk needs 'software_raid' controller", but does that mean _all_ the disks? or _all the software RAID_ disks? We're trying to determine what happens if both RAIDs are given.

Reply:
I think that mean all the disks, beacause the different of `software raid` and `hardware raid` is `controller`,and if only include software RAID disks the check is meaningless.

So i think if we're trying to determine `software raid` and `hardware raid` at the same time, an error will be reported.

Reply:
someone from Ironic community told me they can both be configured if "done carefully". That looks to me like a complication, and we should not have this in our implementation. I think it is sufficient to document it as a recommendation for the user to _not_ apply them both at the same time, since it can become complicated.

However, if we decide that we need to restrict users from using one at a time, we can do that as well by reporting an error to user, without passing it onto Ironic. As @dhellmann said, BMO should be considered as something independent of Ironic.

Reply:
I am not sure but if we configure `software raid` in a `manual clean` and configure `hardware raid` in another `manual clean`, is it possible to configure software RAID and hardware RAID on a same node?

Reply:
The root disk must be RAID-1 because otherwise Grub will need to understand how to assemble mdraid to be able to boot anything. RAID-1 is safe in this regard because any mirror looks just like a normal disk.

Reply:
@Hellcatlk Let's say for now, we aren't going to allow this, since it will lead to unnecessary complications. Let's "discourage" this in code for now, and document this. If a need comes up later to have both, a patch can be developed accordingly.

@dtantsur I see. In that case, we can stick to RAID-1 on the root disk.

Reply:
@dtantsur That's an exceptionally clear explanation for the requirement. I hope that's in some documentation somewhere. :-)

I will add that detail to this design doc before we merge it.